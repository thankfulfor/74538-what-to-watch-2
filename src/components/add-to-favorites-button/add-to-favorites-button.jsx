import * as React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';

import {Operation} from '../../operations/operation.js';
import {URLS} from '../../utils/constants.js';

export const AddToFavoritesButton = (props) => {
  const {filmId, isFavorite, onFavoriteButtonClick, isLoggedIn, history} = props;

  const loginPageUrl = URLS.LOGIN_PAGE_URL;

  const favoriteButtonClickHandler = () => {
    if (isLoggedIn) {
      onFavoriteButtonClick(filmId, +!isFavorite);
    } else {
      history.push(loginPageUrl);
    }
  };

  return (
    <React.Fragment>
      <div className="visually-hidden">
        <svg>
          <symbol id="in-list" viewBox="0 0 18 14">
            <path fillRule="evenodd" clipRule="evenodd" d="M2.40513 5.35353L6.1818 8.90902L15.5807 0L18 2.80485L6.18935 14L0 8.17346L2.40513 5.35353Z" fill="#EEE5B5"/>
          </symbol>
        </svg>
      </div>
      <button
        className="btn btn--list movie-card__button"
        type="button"
        onClick={favoriteButtonClickHandler}
      >
        <svg viewBox="0 0 19 20" width="19" height="20">
          <use xlinkHref={isFavorite ? `#in-list` : `#add`} />
        </svg>
        <span>My list</span>
      </button>
    </React.Fragment>
  );
};

AddToFavoritesButton.propTypes = {
  isFavorite: PropTypes.bool.isRequired,
  filmId: PropTypes.number.isRequired,
  onFavoriteButtonClick: PropTypes.func.isRequired,
  isLoggedIn: PropTypes.bool.isRequired,
  history: PropTypes.object.isRequired,
};

const mapStateToProps = (state, ownProps) => {
  return Object.assign({}, ownProps, {
    isLoggedIn: state.isLoggedIn,
    isFavorite: state.isFavorite
  });
};

const mapDispatchToProps = (dispatch) => {
  return ({
    onFavoriteButtonClick: (filmId, status) => {
      dispatch(Operation.setFavorite(filmId, status));
    }
  });
};

export default connect(mapStateToProps, mapDispatchToProps)(AddToFavoritesButton);
