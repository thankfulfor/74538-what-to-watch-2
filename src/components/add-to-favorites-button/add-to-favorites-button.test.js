import React from 'react';
import {AddToFavoritesButton} from './add-to-favorites-button.jsx';
import Enzyme, {mount} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import {BrowserRouter} from 'react-router-dom';

Enzyme.configure({adapter: new Adapter()});

it(`AddToFavoritesButton корректно рендерится после перезапуска`, () => {
  const clickHandler = function () {};

  const tree = mount(
      <BrowserRouter>
        <AddToFavoritesButton
          filmId={0}
          isFavorite={false}
          isLoggedIn={false}
          onFavoriteButtonClick={clickHandler}
        />
      </BrowserRouter>);
  expect(toJson(tree)).toMatchSnapshot();
});
