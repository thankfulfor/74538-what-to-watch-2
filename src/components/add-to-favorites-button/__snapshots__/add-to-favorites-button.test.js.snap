// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddToFavoritesButton корректно рендерится после перезапуска 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <AddToFavoritesButton
      filmId={0}
      isFavorite={false}
      isLoggedIn={false}
      onFavoriteButtonClick={[Function]}
    >
      <div
        className="visually-hidden"
      >
        <svg>
          <symbol
            id="in-list"
            viewBox="0 0 18 14"
          >
            <path
              clipRule="evenodd"
              d="M2.40513 5.35353L6.1818 8.90902L15.5807 0L18 2.80485L6.18935 14L0 8.17346L2.40513 5.35353Z"
              fill="#EEE5B5"
              fillRule="evenodd"
            />
          </symbol>
        </svg>
      </div>
      <button
        className="btn btn--list movie-card__button"
        onClick={[Function]}
        type="button"
      >
        <svg
          height="20"
          viewBox="0 0 19 20"
          width="19"
        >
          <use
            xlinkHref="#add"
          />
        </svg>
        <span>
          My list
        </span>
      </button>
    </AddToFavoritesButton>
  </Router>
</BrowserRouter>
`;
